File: src/tmux.ts
/** Typed wrappers around tmux commands. */

import { exec } from "./subprocess.ts";

function run(args: string[]) {
  return exec("tmux", args);
}

export async function hasSession(name: string): Promise<boolean> {
  const { success } = await run(["has-session", "-t", name]);
  return success;
}

export async function createSession(name: string): Promise<void> {
  const { success, stderr } = await run(["new-session", "-d", "-s", name]);
  if (!success) throw new Error(`tmux new-session failed: ${stderr}`);
}

export async function killSession(name: string): Promise<void> {
  const { success, stderr } = await run(["kill-session", "-t", name]);
  if (!success) throw new Error(`tmux kill-session failed: ${stderr}`);
}

export async function createWindow(
  session: string,
  name: string,
): Promise<void> {
  const { success, stderr } = await run([
    "new-window",
    "-t",
    session,
    "-n",
    name,
  ]);
  if (!success) throw new Error(`tmux new-window failed: ${stderr}`);
}

export async function sendKeys(
  target: string,
  text: string,
  enter = true,
): Promise<void> {
  // Send text and Enter as separate calls -- TUI apps (Codex, Gemini) don't
  // reliably submit when Enter is appended to the same send-keys invocation.
  if (text) {
    const { success, stderr } = await run(["send-keys", "-t", target, text]);
    if (!success) throw new Error(`tmux send-keys failed: ${stderr}`);
  }
  if (enter) {
    const { success, stderr } = await run([
      "send-keys",
      "-t",
      target,
      "Enter",
    ]);
    if (!success) throw new Error(`tmux send-keys (enter) failed: ${stderr}`);
  }
}

export async function capturePane(
  target: string,
  lines = 50,
): Promise<string> {
  const { success, stdout, stderr } = await run([
    "capture-pane",
    "-t",
    target,
    "-p",
    "-S",
    `-${lines}`,
  ]);
  if (!success) throw new Error(`tmux capture-pane failed: ${stderr}`);
  return stdout;
}

export interface WindowInfo {
  name: string;
  active: boolean;
  index: number;
}

export async function listWindows(session: string): Promise<WindowInfo[]> {
  const { success, stdout, stderr } = await run([
    "list-windows",
    "-t",
    session,
    "-F",
    "#{window_index}\t#{window_name}\t#{window_active}",
  ]);
  if (!success) throw new Error(`tmux list-windows failed: ${stderr}`);
  if (!stdout) return [];
  return stdout.split("\n").map((line) => {
    const [index, name, active] = line.split("\t");
    return { index: parseInt(index), name, active: active === "1" };
  });
}

export interface PaneInfo {
  windowName: string;
  paneIndex: number;
  paneId: string; // stable %N id
  paneDead: boolean;
  currentCommand: string;
}

export async function listPanes(session: string): Promise<PaneInfo[]> {
  const { success, stdout, stderr } = await run([
    "list-panes",
    "-t",
    session,
    "-s",
    "-F",
    "#{window_name}\t#{pane_index}\t#{pane_dead}\t#{pane_current_command}\t#{pane_id}",
  ]);
  if (!success) throw new Error(`tmux list-panes failed: ${stderr}`);
  if (!stdout) return [];
  return stdout.split("\n").map((line) => {
    const [windowName, idx, dead, cmd, paneId] = line.split("\t");
    return {
      windowName,
      paneIndex: parseInt(idx),
      paneId: paneId ?? "",
      paneDead: dead === "1",
      currentCommand: cmd ?? "",
    };
  });
}

/** Get session creation time as a unix epoch (seconds), or null. */
export async function sessionCreated(name: string): Promise<number | null> {
  const { success, stdout } = await run([
    "display-message",
    "-t",
    name,
    "-p",
    "#{session_created}",
  ]);
  if (!success || !stdout) return null;
  const epoch = parseInt(stdout.trim());
  return isNaN(epoch) ? null : epoch;
}

export async function selectWindow(
  session: string,
  name: string,
): Promise<void> {
  const { success, stderr } = await run([
    "select-window",
    "-t",
    `${session}:${name}`,
  ]);
  if (!success) throw new Error(`tmux select-window failed: ${stderr}`);
}

/** Switch the current tmux client to a different session. */
export async function switchClient(targetSession: string): Promise<void> {
  const { success, stderr } = await run([
    "switch-client",
    "-t",
    targetSession,
  ]);
  if (!success) throw new Error(`tmux switch-client failed: ${stderr}`);
}

/** Swap two panes. Use fully-qualified targets (e.g. session:window.pane). */
export async function swapPane(
  source: string,
  target: string,
): Promise<void> {
  const { success, stderr } = await run([
    "swap-pane",
    "-s",
    source,
    "-t",
    target,
  ]);
  if (!success) throw new Error(`tmux swap-pane failed: ${stderr}`);
}

/** Display a message on the tmux status line. */
export async function displayMessage(
  session: string,
  message: string,
): Promise<void> {
  const { success, stderr } = await run([
    "display-message",
    "-t",
    session,
    message,
  ]);
  if (!success) throw new Error(`tmux display-message failed: ${stderr}`);
}

/** Split the target window horizontally, creating a new pane below. */
export async function splitWindow(
  session: string,
  targetWindow: string,
  percentage?: number,
): Promise<void> {
  const args = [
    "split-window",
    "-t",
    `${session}:${targetWindow}`,
    "-v",
  ];
  if (percentage !== undefined) {
    args.push("-p", String(percentage));
  }
  const { success, stderr } = await run(args);
  if (!success) throw new Error(`tmux split-window failed: ${stderr}`);
}

/** Set a user option on a specific pane. */
export async function setPaneOption(
  target: string,
  option: string,
  value: string,
): Promise<void> {
  const { success, stderr } = await run([
    "set-option",
    "-p",
    "-t",
    target,
    option,
    value,
  ]);
  if (!success) throw new Error(`tmux set-option failed: ${stderr}`);
}

/** Find a pane's %N ID by its @jackops_role user option within a session. */
export async function findPaneByRole(
  session: string,
  role: string,
): Promise<string | null> {
  const { success, stdout } = await run([
    "list-panes",
    "-s",
    "-t",
    session,
    "-F",
    "#{pane_id}\t#{@jackops_role}",
  ]);
  if (!success || !stdout) return null;
  for (const line of stdout.split("\n")) {
    const [paneId, paneRole] = line.split("\t");
    if (paneRole === role) return paneId;
  }
  return null;
}

/** Rename the first window (index 0) created with the session. */
export async function renameWindow(
  session: string,
  index: number,
  name: string,
): Promise<void> {
  const { success, stderr } = await run([
    "rename-window",
    "-t",
    `${session}:${index}`,
    name,
  ]);
  if (!success) throw new Error(`tmux rename-window failed: ${stderr}`);
}

File: src/tmux/tmux.go
package tmux

import (
	"context"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

const cmdTimeout = 10 * time.Second

func run(args ...string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "tmux", args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("tmux %s: %w: %s", args[0], err, strings.TrimSpace(string(out)))
	}
	return strings.TrimRight(string(out), "\n"), nil
}

func HasSession(name string) bool {
	_, err := run("has-session", "-t", name)
	return err == nil
}

func CreateSession(name string) error {
	_, err := run("new-session", "-d", "-s", name)
	return err
}

func KillSession(name string) error {
	_, err := run("kill-session", "-t", name)
	return err
}

func CreateWindow(session, name string) error {
	_, err := run("new-window", "-t", session, "-n", name)
	return err
}

func CreateWindowDetached(session, name string) error {
	_, err := run("new-window", "-d", "-t", session, "-n", name)
	return err
}

func KillWindow(session, name string) error {
	_, err := run("kill-window", "-t", fmt.Sprintf("%s:%s", session, name))
	return err
}

// KillPane kills a specific pane by its ID (e.g. "%5").
func KillPane(paneID string) error {
	_, err := run("kill-pane", "-t", paneID)
	return err
}

func RenameWindow(session string, index int, name string) error {
	target := fmt.Sprintf("%s:%d", session, index)
	_, err := run("rename-window", "-t", target, name)
	return err
}

func SelectWindow(session, name string) error {
	target := fmt.Sprintf("%s:%s", session, name)
	_, err := run("select-window", "-t", target)
	return err
}

func SplitWindow(session, targetWindow string, percentage int) error {
	args := []string{"split-window", "-t", fmt.Sprintf("%s:%s", session, targetWindow), "-v"}
	if percentage > 0 {
		args = append(args, "-p", strconv.Itoa(percentage))
	}
	_, err := run(args...)
	return err
}

// SendKeys sends text to a tmux pane. Text is prefixed with a space to
// keep commands out of shell history. Text and Enter are separate calls
// for TUI app compatibility.
func SendKeys(target, text string, enter bool) error {
	if text != "" {
		if _, err := run("send-keys", "-t", target, " "+text); err != nil {
			return err
		}
	}
	if enter {
		if _, err := run("send-keys", "-t", target, "Enter"); err != nil {
			return err
		}
	}
	return nil
}

// SendKeysDoubleEnter sends text followed by two Enter presses with a short
// delay between them. Claude Code requires a second Enter to submit a prompt.
func SendKeysDoubleEnter(target, text string) error {
	if text != "" {
		if _, err := run("send-keys", "-t", target, " "+text); err != nil {
			return err
		}
	}
	if _, err := run("send-keys", "-t", target, "Enter"); err != nil {
		return err
	}
	time.Sleep(500 * time.Millisecond)
	_, err := run("send-keys", "-t", target, "Enter")
	return err
}

func CapturePane(target string, lines int) (string, error) {
	return run("capture-pane", "-t", target, "-p", "-S", fmt.Sprintf("-%d", lines))
}

type WindowInfo struct {
	Name   string
	Index  int
	Active bool
}

func ListWindows(session string) ([]WindowInfo, error) {
	out, err := run("list-windows", "-t", session, "-F", "#{window_index}\t#{window_name}\t#{window_active}")
	if err != nil {
		return nil, err
	}
	if out == "" {
		return nil, nil
	}
	var windows []WindowInfo
	for _, line := range strings.Split(out, "\n") {
		parts := strings.SplitN(line, "\t", 3)
		if len(parts) < 3 {
			continue
		}
		idx, _ := strconv.Atoi(parts[0])
		windows = append(windows, WindowInfo{
			Index:  idx,
			Name:   parts[1],
			Active: parts[2] == "1",
		})
	}
	return windows, nil
}

type PaneInfo struct {
	WindowName     string
	PaneIndex      int
	PaneDead       bool
	CurrentCommand string
}

func ListPanes(session string) ([]PaneInfo, error) {
	out, err := run("list-panes", "-t", session, "-a", "-F", "#{window_name}\t#{pane_index}\t#{pane_dead}\t#{pane_current_command}")
	if err != nil {
		return nil, err
	}
	if out == "" {
		return nil, nil
	}
	var panes []PaneInfo
	for _, line := range strings.Split(out, "\n") {
		parts := strings.SplitN(line, "\t", 4)
		if len(parts) < 4 {
			continue
		}
		idx, _ := strconv.Atoi(parts[1])
		panes = append(panes, PaneInfo{
			WindowName:     parts[0],
			PaneIndex:      idx,
			PaneDead:       parts[2] == "1",
			CurrentCommand: parts[3],
		})
	}
	return panes, nil
}

func DisplayMessage(session, message string) error {
	_, err := run("display-message", "-t", session, message)
	return err
}

func SwitchClient(targetSession string) error {
	_, err := run("switch-client", "-t", targetSession)
	return err
}

func SessionCreated(name string) (int64, error) {
	out, err := run("display-message", "-t", name, "-p", "#{session_created}")
	if err != nil {
		return 0, err
	}
	epoch, err := strconv.ParseInt(strings.TrimSpace(out), 10, 64)
	if err != nil {
		return 0, fmt.Errorf("parsing session_created: %w", err)
	}
	return epoch, nil
}

// SetPaneOption sets a user option on a tmux pane (e.g. @jackin_role).
func SetPaneOption(target, key, value string) error {
	_, err := run("set-option", "-p", "-t", target, key, value)
	return err
}

// FindPaneByRole finds a pane with a matching @jackin_role user option.
// Only matches panes in the specified session to avoid cross-session collisions.
func FindPaneByRole(session, role string) (string, error) {
	out, err := run("list-panes", "-t", session, "-a", "-F", "#{session_name}\t#{pane_id}\t#{@jackin_role}")
	if err != nil {
		return "", err
	}
	for _, line := range strings.Split(out, "\n") {
		parts := strings.SplitN(line, "\t", 3)
		if len(parts) == 3 && parts[0] == session && parts[2] == role {
			return parts[1], nil
		}
	}
	return "", nil
}

// SwapPane swaps two panes by their IDs.
func SwapPane(src, dst string) error {
	_, err := run("swap-pane", "-s", src, "-t", dst)
	return err
}

// PopupOpen opens a tmux popup overlay running the given shell command.
// It blocks until the popup closes (command exits or Escape is pressed).
func PopupOpen(session, shellCmd string, widthPct, heightPct int) error {
	_, err := run("display-popup", "-t", session,
		"-E", "-w", fmt.Sprintf("%d%%", widthPct), "-h", fmt.Sprintf("%d%%", heightPct),
		shellCmd)
	return err
}

// PaneTarget resolves a @jackin_role to a stable pane_id target.
// Falls back to session:role (window-based) if the role is not found.
func PaneTarget(session, role string) string {
	paneID, err := FindPaneByRole(session, role)
	if err != nil || paneID == "" {
		return session + ":" + role
	}
	return paneID
}
